### ISSUE TITLE
[FEATURE] [GrootN1.6] Full-Iteration / Per-Microbatch CUDA Graph Support for GRooT N1.6 Training

### ISSUE THREAD

**Issue by Dana57740** [baidu-baige/LoongForge#49]
# Full-Iteration / Per-Microbatch CUDA Graph Support for GRooT N1.6 Training

## Summary

Two CUDA graph capture modes for GRooT N1.6 VLA training, both eliminating kernel-launch and Python-interpreter overhead:

- **`per_microbatch`** *(recommended)* — one sub-graph per microbatch with RNG (`Beta.sample` / `torch.randn`) externalized to host between replays. Loss tracks pure-eager bit-near-exactly.
- **`full_iteration`** — single graph captures one entire iteration (all microbatches' fwd + bwd + grad-sync). Slightly faster, but RNG is baked into the graph and loss does not fully align with pure-eager.

**Use `per_microbatch` by default.** `full_iteration` is kept for benchmarking / launch-overhead-bound scenarios where loss alignment is not required.

## Motivation

Profiling showed 32.3% of GPU kernel time spent on small elementwise ops with high launch overhead. CUDA graph amortizes this. `full_iteration` is fastest but its baked RNG diverges from pure-eager loss (~3.47% RelErr); `per_microbatch` recovers ~0.81% RelErr at essentially the same throughput, and is therefore the recommended mode.

## Results (8×A800, mb=16, gbs=128, 100 iters, w/ DDP overlap)

| Mode | Iter time | Throughput (token/s/GPU) | Speedup | Avg loss RelErr vs pure-eager |
| --- | --- | --- | --- | --- |
| pure_eager | 478.6 ms | 8 676 | 1.00× | — |
| full_iteration | 330.3 ms | 13 130 | **1.51×** | 3.47% (max 12.38%) |
| **per_microbatch** | 333.8 ms | 12 991 | **1.50×** | **0.81%** (max 4.35%) |

per_microbatch matches full_iteration in throughput, with ~4× better loss alignment.

## Implementation

Three-phase state machine shared by both modes: **Warmup → Capture → Replay**. Warmup runs N eager iterations to populate non-capturable caches (`tolist()`, resized positional embeddings, etc.). Capture records into `torch.cuda.CUDAGraph` with static buffers. Replay swaps new data via `copy_()` and replays; `StaticBufferLoader` self-heals on buffer realloc / shape change.

### Key components

| File | Change |
| --- | --- |
| `cuda_graph_base.py` *(new)* | `BaseCudaGraphWrapper` template-method base — warmup/capture/replay state machine, `StaticBufferLoader`, graph-invalidation self-heal, loss-allreduce suppression hooks. Shared by both modes. |
| `cuda_graph_config.py` *(new)* | Centralized mode-query interface (`is_full_iteration_graph`, `is_per_microbatch_graph`, …). |
| `groot_graph_mixin.py` *(new)* | Shared `_suppress_loss_allreduce()` contextmanager. |
| `groot_full_iteration_graph.py` *(new)* | Single graph capturing one full training iteration; baked NCCL allreduce; loss allreduce deferred to post-replay. |
| `groot_per_microbatch_graph.py` *(new)* | Per-MB sub-graphs + DDP-overlap integration: first N-1 captured under `_suppress_grad_sync` (no NCCL); last captured outside, baking NCCL kernels into the graph so replay preserves overlap. |
| `modeling_eagle3_vl.py` | Graph-safe FA2 attention: `argsort(stable=True)` replaces non-capturable `nonzero`; deferred monkey-patch of `_flash_attention_forward`. |
| `modeling_siglip2.py` | Caches spatial shapes / position embeddings / window sort indices (`copy_()` in-place updates only — assignment kills graph addresses, was a real bug). `antialias=True` disabled during capture. |
| `modeling_groot.py` | Pre-allocated padding buffers via `copy_()` (no `torch.zeros` / `torch.cat` during capture). `Beta.sample` → per_microbatch: eager; full_iteration: inverse-CDF when β=1. |
| `processor_groot.py` | `--cuda-graph-pad-length` for fixed-shape batches. |
| `training_utils.py` | Registry-based wrapper resolution (`register_full_iteration_graph_wrapper` / `register_per_microbatch_graph_wrapper`); falls back to Megatron's built-in `FullCudaGraphWrapper` when no model-specific wrapper is registered (full_iteration only). |
| `sft_groot.py` | Deferred loss allreduce, CPU `total_inputs`, reproducible DataLoader seeding; registers both wrappers. |
| `arguments.py` | Adds `per_microbatch` choice to `--cuda-graph-scope`; new `--cuda-graph-pad-length`; lazy import fix. |

13 files changed, +1876 / −143.

## Configuration

CUDA graph is opt-in (the example script ships with it disabled). To enable, append:

```bash
--cuda-graph-impl local
--cuda-graph-scope per_microbatch    # recommended; use full_iteration only when loss alignment isn't required
--cuda-graph-warmup-steps 3
--cuda-graph-pad-length 220
--no-check-for-nan-in-loss-and-grad
```

## Limitations

- `--cuda-graph-pad-length` is required (fixed-shape batches); over-length sequences are truncated with a warning.
- `full_iteration` does not support `noise_beta_beta != 1` (no closed-form inverse CDF for general Beta) — use `per_microbatch`.
- SFT only; pretrain integration is out of scope.
- DDP overlap baking in `per_microbatch` assumes the last microbatch is in the same forward_backward call; pipeline parallelism is not yet validated end-to-end.
